File: frontend/src/prediction/recipeParser.js
import { analyzeCookie, NESTLE_TOLL_HOUSE } from './scienceEngine.js';

const EMPTY_RECIPE = {
  ...NESTLE_TOLL_HOUSE,
  flour_g: 0,
  butter_g: 0,
  shortening_g: 0,
  oil_g: 0,
  white_sugar_g: 0,
  light_brown_sugar_g: 0,
  dark_brown_sugar_g: 0,
  egg_g: 0,
  egg_yolk_g: 0,
  baking_soda_g: 0,
  baking_powder_g: 0,
  cornstarch_g: 0,
  chocolate_g: 0,
};

const CUP_WEIGHTS = {
  flour_g: 125,
  butter_g: 226,
  shortening_g: 205,
  oil_g: 218,
  white_sugar_g: 200,
  light_brown_sugar_g: 220,
  baking_soda_g: 220.8,
  baking_powder_g: 192,
  cornstarch_g: 128,
  chocolate_g: 170,
};

const FRACTIONS = {
  '¼': '1/4',
  '½': '1/2',
  '¾': '3/4',
  '⅓': '1/3',
  '⅔': '2/3',
  '⅛': '1/8',
  '⅜': '3/8',
  '⅝': '5/8',
  '⅞': '7/8',
};

const UNIT_ALIASES = {
  cup: 'cup', cups: 'cup', c: 'cup',
  tablespoon: 'tbsp', tablespoons: 'tbsp', tbsp: 'tbsp', tbs: 'tbsp',
  teaspoon: 'tsp', teaspoons: 'tsp', tsp: 'tsp',
  gram: 'g', grams: 'g', g: 'g', kilogram: 'kg', kilograms: 'kg', kg: 'kg',
  ounce: 'oz', ounces: 'oz', oz: 'oz', pound: 'lb', pounds: 'lb', lb: 'lb', lbs: 'lb',
  milliliter: 'ml', milliliters: 'ml', ml: 'ml', stick: 'stick', sticks: 'stick',
};

const normalizeFractions = (text) => {
  let normalized = text.replaceAll('\u00a0', ' ').replaceAll('⁄', '/');
  Object.entries(FRACTIONS).forEach(([glyph, fraction]) => {
    normalized = normalized.replace(new RegExp(`(\\d)${glyph}`, 'g'), `$1 ${fraction}`);
    normalized = normalized.replaceAll(glyph, fraction);
  });
  return normalized;
};

const parseNumber = (text) => text.trim().split(/\s+/).reduce((total, part) => {
  if (!part.includes('/')) return total + Number(part);
  const [numerator, denominator] = part.split('/').map(Number);
  return total + numerator / denominator;
}, 0);

const quantityAndUnit = (line) => {
  const match = line.match(/^\s*(\d+\s+\d+\/\d+|\d+\/\d+|\d+(?:\.\d+)?)\s*(.*)$/i);
  if (!match) return null;
  const quantity = parseNumber(match[1]);
  const remainder = match[2].trim();
  const packageMatch = remainder.match(/^\(?\s*(\d+(?:\.\d+)?)\s*[- ]?\s*(ounce|ounces|oz|g|grams?)\s*\)?\s+(.*)$/i);
  if (packageMatch) {
    return [quantity * Number(packageMatch[1]), UNIT_ALIASES[packageMatch[2].toLowerCase()], packageMatch[3].trim()];
  }
  const unitMatch = remainder.match(/^([a-zA-Z]+)\.?\b\s*(.*)$/);
  if (unitMatch && UNIT_ALIASES[unitMatch[1].toLowerCase()]) {
    return [quantity, UNIT_ALIASES[unitMatch[1].toLowerCase()], unitMatch[2].trim()];
  }
  return [quantity, 'count', remainder];
};

const ingredientKey = (line) => {
  const value = line.toLowerCase();
  if (/\bbaking\s+soda\b|\bbicarbonate\s+of\s+soda\b/.test(value)) return 'baking_soda_g';
  if (/\bbaking\s+powder\b/.test(value)) return 'baking_powder_g';
  if (/\bcorn\s*starch\b|\bcornflour\b/.test(value)) return 'cornstarch_g';
  if (/\b(chocolate\s+chips?|chocolate\s+morsels?|chopped\s+chocolate)\b/.test(value)) return 'chocolate_g';
  if (/\begg\s+yolks?\b|\byolks?\b/.test(value)) return 'egg_yolk_g';
  if (/\beggs?\b/.test(value)) return 'egg_g';
  if (/\b(light|dark)?\s*brown\s+sugar\b/.test(value)) return 'light_brown_sugar_g';
  if (/\b(granulated|white|caster|superfine)\s+sugar\b/.test(value)) return 'white_sugar_g';
  if (/\bshortening\b/.test(value)) return 'shortening_g';
  if (/\b(vegetable|canola|coconut)?\s*oil\b/.test(value)) return 'oil_g';
  if (/\bbutter\b/.test(value)) return 'butter_g';
  if (/\b(all[- ]purpose|plain)?\s*flour\b/.test(value)) return 'flour_g';
  if (/\bsugar\b/.test(value) && !value.includes('brown')) return 'white_sugar_g';
  return null;
};

const toGrams = (key, quantity, unit) => {
  if (unit === 'g') return quantity;
  if (unit === 'kg') return quantity * 1000;
  if (unit === 'oz') return quantity * 28.3495;
  if (unit === 'lb') return quantity * 453.592;
  if (key === 'egg_g' && unit === 'count') return quantity * 50;
  if (key === 'egg_yolk_g' && unit === 'count') return quantity * 18;
  if (key === 'butter_g' && unit === 'stick') return quantity * 113;
  const cupWeight = CUP_WEIGHTS[key];
  if (!cupWeight) return null;
  if (unit === 'cup') return quantity * cupWeight;
  if (unit === 'tbsp') return quantity * cupWeight / 16;
  if (unit === 'tsp') return quantity * cupWeight / 48;
  if (unit === 'ml') return quantity * cupWeight / 236.588;
  return null;
};

const detectButterState = (text) => {
  const value = text.toLowerCase();
  if (/\bbrown(?:ed)?\s+butter\b|\bbutter[^\n,.]{0,24}\bbrown(?:ed)?\b/.test(value)) return ['browned', true];
  if (/\bmelted\s+butter\b|\bbutter[^\n,.]{0,24}\bmelted\b/.test(value)) return ['melted', true];
  if (/\b(cold|chilled|frozen)\s+butter\b|\bbutter[^\n,.]{0,24}\b(cold|chilled|frozen)\b/.test(value)) return ['cold', true];
  if (/\bsoftened\s+butter\b|\bbutter[^\n,.]{0,24}\b(softened|room temperature)\b/.test(value)) return ['softened', true];
  return ['softened', false];
};

const detectMixing = (text, butterState) => {
  const value = text.toLowerCase();
  if (/\bcream(?:ed|ing)?\b|\bbeat[^.\n]{0,60}\bbutter\b[^.\n]{0,60}\bsugar/.test(value)) return ['creamed', true];
  if (/\b(stir|stirred|whisk|whisked|mix|mixed)\b/.test(value)) return ['stirred', true];
  return [['melted', 'browned'].includes(butterState) ? 'stirred' : 'creamed', false];
};

const detectChill = (text) => {
  const value = text.toLowerCase();
  if (/\b(no|without)\s+chill/.test(value)) return [0, true];
  if (/\b(chill|refrigerat\w*)[^.\n]{0,50}\bovernight\b/.test(value)) return [12, true];
  const match = value.match(/\b(?:chill|refrigerat\w*)[^.\n]{0,50}?(\d+(?:\.\d+)?|\d+\/\d+)\s*(hours?|hrs?|minutes?|mins?)\b/);
  if (!match) return [0, false];
  const duration = parseNumber(match[1]) / (match[2].startsWith('min') ? 60 : 1);
  return [Math.round(duration * 100) / 100, true];
};

const detectTemperature = (text) => {
  const match = text.toLowerCase().match(/\b(?:preheat|oven|bake)[^.\n]{0,60}?(\d{2,3})\s*°?\s*([fc])\b/)
    || text.toLowerCase().match(/\b(\d{2,3})\s*°\s*([fc])\b/);
  if (!match) return [350, false];
  const temperature = match[2] === 'c' ? Number(match[1]) * 9 / 5 + 32 : Number(match[1]);
  return [Math.round(temperature), true];
};

const detectBakeTime = (text) => {
  const value = text.toLowerCase().replace(/[–—]/g, '-');
  const range = value.match(/\bbake[^.\n]{0,90}?(\d+(?:\.\d+)?)\s*(?:-|to)\s*(\d+(?:\.\d+)?)\s*(?:minutes?|mins?)\b/);
  if (range) return [(Number(range[1]) + Number(range[2])) / 2, true];
  const match = value.match(/\bbake[^.\n]{0,90}?(\d+(?:\.\d+)?)\s*(?:minutes?|mins?)\b/);
  return match ? [Number(match[1]), true] : [10, false];
};

export function parseCookieRecipe(recipeText) {
  if (!recipeText?.trim()) throw new Error('Recipe text cannot be empty.');
  const normalizedText = normalizeFractions(recipeText);
  const recipe = { ...EMPTY_RECIPE };
  const warnings = [];
  const unparsedLines = [];

  normalizedText.split(/[\n;]+/).forEach((rawLine) => {
    const line = rawLine.replace(/^\s*[-*•]\s*/, '').trim();
    if (!line) return;
    const eggCombo = line.match(/^\s*(\d+(?:\.\d+)?)\s+(?:large\s+)?eggs?\s*(?:\+|plus|and)\s*(\d+(?:\.\d+)?)\s+(?:large\s+)?(?:egg\s+)?yolks?\b/i);
    if (eggCombo) {
      recipe.egg_g += Number(eggCombo[1]) * 50;
      recipe.egg_yolk_g += Number(eggCombo[2]) * 18;
      return;
    }
    const amount = quantityAndUnit(line);
    if (!amount) return;
    const [quantity, unit, remainder] = amount;
    const key = ingredientKey(remainder);
    if (!key) {
      if (!/\b(salt|vanilla(?:\s+extract)?)\b/i.test(remainder) && unit !== 'count') unparsedLines.push(rawLine.trim());
      return;
    }
    const grams = toGrams(key, quantity, unit);
    if (grams == null) unparsedLines.push(rawLine.trim());
    else recipe[key] = Math.round((recipe[key] + grams) * 100) / 100;
  });

  const [butterState, butterFound] = detectButterState(normalizedText);
  const [mixingMethod, mixingFound] = detectMixing(normalizedText, butterState);
  const [chillHours, chillFound] = detectChill(normalizedText);
  const [bakeTemp, temperatureFound] = detectTemperature(normalizedText);
  const [bakeTime, bakeTimeFound] = detectBakeTime(normalizedText);
  const sizeMatch = normalizedText.toLowerCase().match(/\b(?:portion|scoop|dough balls?|cookies?)[^.\n]{0,45}?(\d+(?:\.\d+)?)\s*g\b/);

  Object.assign(recipe, {
    butter_state: butterState,
    mixing_method: mixingMethod,
    chill_hours: chillHours,
    dough_temperature: chillHours > 0 ? 'chilled' : 'room',
    bake_temp_f: bakeTemp,
    bake_time_min: bakeTime,
    cookie_size_g: sizeMatch ? Number(sizeMatch[1]) : 30,
  });

  if (recipe.flour_g === 0) warnings.push('No flour quantity could be parsed.');
  if (recipe.butter_g + recipe.shortening_g + recipe.oil_g === 0) warnings.push('No butter, shortening, or oil quantity could be parsed.');
  if (recipe.white_sugar_g + recipe.light_brown_sugar_g === 0) warnings.push('No white or brown sugar quantity could be parsed.');
  if (recipe.chocolate_g === 0) warnings.push('No chocolate quantity was found; the prediction describes the base cookie dough.');
  if (recipe.butter_g > 0 && !butterFound) warnings.push('No butter state was found; defaulted to softened butter.');
  if (!mixingFound) warnings.push(`No mixing method was found; defaulted to ${recipe.mixing_method} mixing.`);
  if (!chillFound) warnings.push('No chill instruction was found; defaulted to 0 hours.');
  if (!temperatureFound) warnings.push('No bake temperature was found; defaulted to 350°F.');
  if (!bakeTimeFound) warnings.push('No bake time was found; defaulted to 10 minutes.');
  if (!sizeMatch) warnings.push('No cookie size was found; defaulted to 30 g.');
  if (unparsedLines.length > 0) warnings.push(`Could not map these measured ingredient lines: ${unparsedLines.slice(0, 4).join('; ')}`);

  return { recipe, warnings, unparsed_lines: unparsedLines };
}

export function analyzeRecipeText(recipeText) {
  const parsed = parseCookieRecipe(recipeText);
  const analysis = analyzeCookie(parsed.recipe);
  const warnings = [...new Set([
    ...parsed.warnings,
    ...(analysis.warnings || []),
    ...(analysis.cookie_failed ? analysis.reason || [] : []),
  ])];
  return {
    parsed_recipe: parsed.recipe,
    prediction: analysis.prediction,
    ml_prediction: analysis.ml_prediction,
    confidence: analysis.confidence,
    warnings,
    explanations: analysis.explanations || [],
    cookie_failed: analysis.cookie_failed,
  };
}

File: src/recipe_parser.py
"""Reliable, rule-based parsing for chocolate chip cookie recipes.

The parser intentionally supports a narrow ingredient vocabulary and common US
baking measurements. It does not infer novel ingredients or call external APIs.
"""

from __future__ import annotations

from dataclasses import dataclass
from fractions import Fraction
import re


DEFAULT_RECIPE = {
    "flour_g": 0.0,
    "butter_g": 0.0,
    "shortening_g": 0.0,
    "oil_g": 0.0,
    "white_sugar_g": 0.0,
    "light_brown_sugar_g": 0.0,
    "egg_g": 0.0,
    "egg_yolk_g": 0.0,
    "baking_soda_g": 0.0,
    "baking_powder_g": 0.0,
    "cornstarch_g": 0.0,
    "chocolate_g": 0.0,
    "butter_state": "softened",
    "mixing_method": "creamed",
    "chill_hours": 0.0,
    "bake_temp_f": 350.0,
    "bake_time_min": 10.0,
    "cookie_size_g": 30.0,
}


# Ingredient-specific cup weights. Tablespoon, teaspoon, and milliliter
# conversions are derived from one US cup = 16 tbsp = 48 tsp = 236.588 ml.
CUP_WEIGHTS_G = {
    "flour_g": 125.0,
    "butter_g": 226.0,
    "shortening_g": 205.0,
    "oil_g": 218.0,
    "white_sugar_g": 200.0,
    "light_brown_sugar_g": 220.0,
    "baking_soda_g": 220.8,  # 4.6 g per teaspoon
    "baking_powder_g": 192.0,  # 4 g per teaspoon
    "cornstarch_g": 128.0,
    "chocolate_g": 170.0,
}


UNICODE_FRACTIONS = {
    "¼": "1/4",
    "½": "1/2",
    "¾": "3/4",
    "⅓": "1/3",
    "⅔": "2/3",
    "⅛": "1/8",
    "⅜": "3/8",
    "⅝": "5/8",
    "⅞": "7/8",
}


UNIT_ALIASES = {
    "cup": "cup",
    "cups": "cup",
    "c": "cup",
    "tablespoon": "tbsp",
    "tablespoons": "tbsp",
    "tbsp": "tbsp",
    "tbs": "tbsp",
    "teaspoon": "tsp",
    "teaspoons": "tsp",
    "tsp": "tsp",
    "gram": "g",
    "grams": "g",
    "g": "g",
    "kilogram": "kg",
    "kilograms": "kg",
    "kg": "kg",
    "ounce": "oz",
    "ounces": "oz",
    "oz": "oz",
    "pound": "lb",
    "pounds": "lb",
    "lb": "lb",
    "lbs": "lb",
    "milliliter": "ml",
    "milliliters": "ml",
    "ml": "ml",
    "stick": "stick",
    "sticks": "stick",
}


@dataclass(frozen=True)
class ParsedCookieRecipe:
    recipe: dict
    warnings: list[str]
    unparsed_lines: list[str]


def _normalize_fractions(text: str) -> str:
    normalized = text.replace("\u00a0", " ").replace("⁄", "/")
    for glyph, fraction in UNICODE_FRACTIONS.items():
        normalized = re.sub(rf"(?<=\d){re.escape(glyph)}", f" {fraction}", normalized)
        normalized = normalized.replace(glyph, fraction)
    return normalized


def _parse_number(value: str) -> float:
    parts = value.strip().split()
    total = 0.0
    for part in parts:
        if "/" in part:
            total += float(Fraction(part))
        else:
            total += float(part)
    return total


def _quantity_and_unit(line: str):
    match = re.match(
        r"^\s*(\d+\s+\d+/\d+|\d+/\d+|\d+(?:\.\d+)?)\s*(.*)$",
        line,
        flags=re.IGNORECASE,
    )
    if not match:
        return None

    quantity = _parse_number(match.group(1))
    remainder = match.group(2).strip()

    # Support package notation such as "1 (12-ounce) bag chocolate chips".
    package_weight = re.match(
        r"^\(?\s*(\d+(?:\.\d+)?)\s*[- ]?\s*(ounce|ounces|oz|g|grams?)\s*\)?\s+",
        remainder,
        flags=re.IGNORECASE,
    )
    if package_weight:
        package_quantity = float(package_weight.group(1))
        package_unit = UNIT_ALIASES[package_weight.group(2).lower()]
        return quantity * package_quantity, package_unit, remainder[package_weight.end():].strip()

    unit_match = re.match(r"^([a-zA-Z]+)\.?\b\s*(.*)$", remainder)
    if unit_match:
        unit_text = unit_match.group(1).lower()
        if unit_text in UNIT_ALIASES:
            return quantity, UNIT_ALIASES[unit_text], unit_match.group(2).strip()

    return quantity, "count", remainder


def _ingredient_key(line: str):
    lowered = line.lower()

    if re.search(r"\bbaking\s+soda\b|\bbicarbonate\s+of\s+soda\b", lowered):
        return "baking_soda_g"
    if re.search(r"\bbaking\s+powder\b", lowered):
        return "baking_powder_g"
    if re.search(r"\bcorn\s*starch\b|\bcornflour\b", lowered):
        return "cornstarch_g"
    if re.search(r"\b(chocolate\s+chips?|chocolate\s+morsels?|chopped\s+chocolate)\b", lowered):
        return "chocolate_g"
    if re.search(r"\begg\s+yolks?\b|\byolks?\b", lowered):
        return "egg_yolk_g"
    if re.search(r"\beggs?\b", lowered):
        return "egg_g"
    if re.search(r"\b(light|dark)?\s*brown\s+sugar\b", lowered):
        return "light_brown_sugar_g"
    if re.search(r"\b(granulated|white|caster|superfine)\s+sugar\b", lowered):
        return "white_sugar_g"
    if re.search(r"\bshortening\b", lowered):
        return "shortening_g"
    if re.search(r"\b(vegetable|canola|coconut)?\s*oil\b", lowered):
        return "oil_g"
    if re.search(r"\bbutter\b", lowered):
        return "butter_g"
    if re.search(r"\b(all[- ]purpose|plain)?\s*flour\b", lowered):
        return "flour_g"
    if re.search(r"\bsugar\b", lowered) and "brown" not in lowered:
        return "white_sugar_g"
    return None


def _convert_to_grams(ingredient_key: str, quantity: float, unit: str):
    if unit == "g":
        return quantity
    if unit == "kg":
        return quantity * 1000
    if unit == "oz":
        return quantity * 28.3495
    if unit == "lb":
        return quantity * 453.592

    if ingredient_key == "egg_g" and unit == "count":
        return quantity * 50
    if ingredient_key == "egg_yolk_g" and unit == "count":
        return quantity * 18
    if ingredient_key == "butter_g" and unit == "stick":
        return quantity * 113

    cup_weight = CUP_WEIGHTS_G.get(ingredient_key)
    if cup_weight is None:
        return None
    if unit == "cup":
        return quantity * cup_weight
    if unit == "tbsp":
        return quantity * cup_weight / 16
    if unit == "tsp":
        return quantity * cup_weight / 48
    if unit == "ml":
        return quantity * cup_weight / 236.588
    return None


def _butter_state(text: str):
    lowered = text.lower()
    if re.search(r"\bbrown(?:ed)?\s+butter\b|\bbutter[^\n,.]{0,24}\bbrown(?:ed)?\b", lowered):
        return "browned", True
    if re.search(r"\bmelted\s+butter\b|\bbutter[^\n,.]{0,24}\bmelted\b", lowered):
        return "melted", True
    if re.search(r"\b(cold|chilled|frozen)\s+butter\b|\bbutter[^\n,.]{0,24}\b(cold|chilled|frozen)\b", lowered):
        return "cold", True
    if re.search(r"\bsoftened\s+butter\b|\bbutter[^\n,.]{0,24}\b(softened|room temperature)\b", lowered):
        return "softened", True
    return "softened", False


def _mixing_method(text: str, butter_state: str):
    lowered = text.lower()
    if re.search(r"\bcream(?:ed|ing)?\b|\bbeat[^.\n]{0,60}\bbutter\b[^.\n]{0,60}\bsugar", lowered):
        return "creamed", True
    if re.search(r"\b(stir|stirred|whisk|whisked|mix|mixed)\b", lowered):
        return "stirred", True
    return ("stirred" if butter_state in {"melted", "browned"} else "creamed"), False


def _chill_hours(text: str):
    lowered = text.lower()
    if re.search(r"\b(no|without)\s+chill", lowered):
        return 0.0, True
    if re.search(r"\b(chill|refrigerat\w*)[^.\n]{0,50}\bovernight\b", lowered):
        return 12.0, True

    match = re.search(
        r"\b(?:chill|refrigerat\w*)[^.\n]{0,50}?(\d+(?:\.\d+)?|\d+/\d+)\s*(hours?|hrs?|minutes?|mins?)\b",
        lowered,
    )
    if not match:
        return 0.0, False
    duration = _parse_number(match.group(1))
    if match.group(2).startswith(("minute", "min")):
        duration /= 60
    return round(duration, 2), True


def _bake_temperature(text: str):
    lowered = text.lower()
    match = re.search(
        r"\b(?:preheat|oven|bake)[^.\n]{0,60}?(\d{2,3})\s*°?\s*([fc])\b",
        lowered,
    )
    if not match:
        match = re.search(r"\b(\d{2,3})\s*°\s*([fc])\b", lowered)
    if not match:
        return 350.0, False

    temperature = float(match.group(1))
    if match.group(2) == "c":
        temperature = temperature * 9 / 5 + 32
    return round(temperature), True


def _bake_time(text: str):
    lowered = text.lower().replace("–", "-").replace("—", "-")
    range_match = re.search(
        r"\bbake[^.\n]{0,90}?(\d+(?:\.\d+)?)\s*(?:-|to)\s*(\d+(?:\.\d+)?)\s*(?:minutes?|mins?)\b",
        lowered,
    )
    if range_match:
        low = float(range_match.group(1))
        high = float(range_match.group(2))
        return round((low + high) / 2, 2), True

    match = re.search(
        r"\bbake[^.\n]{0,90}?(\d+(?:\.\d+)?)\s*(?:minutes?|mins?)\b",
        lowered,
    )
    if match:
        return float(match.group(1)), True
    return 10.0, False


def _cookie_size(text: str):
    lowered = text.lower()
    patterns = [
        r"\b(?:portion|scoop|dough balls?|cookies?)[^.\n]{0,45}?(\d+(?:\.\d+)?)\s*g\b",
        r"\b(\d+(?:\.\d+)?)\s*g\s+(?:portions?|scoops?|dough balls?|cookies?)\b",
    ]
    for pattern in patterns:
        match = re.search(pattern, lowered)
        if match:
            return float(match.group(1)), True
    return 30.0, False


def parse_cookie_recipe(recipe_text: str) -> ParsedCookieRecipe:
    """Parse one chocolate-chip-cookie recipe into the engine's raw schema."""

    if not recipe_text or not recipe_text.strip():
        raise ValueError("Recipe text cannot be empty.")

    normalized_text = _normalize_fractions(recipe_text)
    recipe = dict(DEFAULT_RECIPE)
    warnings = []
    unparsed_lines = []

    raw_lines = re.split(r"[\n;]+", normalized_text)
    for raw_line in raw_lines:
        line = re.sub(r"^\s*[-*•]\s*", "", raw_line).strip()
        if not line:
            continue

        egg_combo = re.match(
            r"^\s*(\d+(?:\.\d+)?)\s+(?:large\s+)?eggs?\s*(?:\+|plus|and)\s*"
            r"(\d+(?:\.\d+)?)\s+(?:large\s+)?(?:egg\s+)?yolks?\b",
            line,
            flags=re.IGNORECASE,
        )
        if egg_combo:
            recipe["egg_g"] += float(egg_combo.group(1)) * 50
            recipe["egg_yolk_g"] += float(egg_combo.group(2)) * 18
            continue

        parsed_amount = _quantity_and_unit(line)
        if not parsed_amount:
            continue
        quantity, unit, remainder = parsed_amount
        ingredient_key = _ingredient_key(remainder)

        if ingredient_key is None:
            # Salt and vanilla do not map to the current prediction schema.
            if re.search(r"\b(salt|vanilla(?:\s+extract)?)\b", remainder.lower()):
                continue
            if unit != "count":
                unparsed_lines.append(raw_line.strip())
            continue

        grams = _convert_to_grams(ingredient_key, quantity, unit)
        if grams is None:
            unparsed_lines.append(raw_line.strip())
            continue
        recipe[ingredient_key] = round(recipe[ingredient_key] + grams, 2)

    recipe["butter_state"], butter_state_found = _butter_state(normalized_text)
    recipe["mixing_method"], mixing_method_found = _mixing_method(
        normalized_text,
        recipe["butter_state"],
    )
    recipe["chill_hours"], chill_found = _chill_hours(normalized_text)
    recipe["bake_temp_f"], temperature_found = _bake_temperature(normalized_text)
    recipe["bake_time_min"], bake_time_found = _bake_time(normalized_text)
    recipe["cookie_size_g"], cookie_size_found = _cookie_size(normalized_text)

    if recipe["flour_g"] == 0:
        warnings.append("No flour quantity could be parsed.")
    if recipe["butter_g"] + recipe["shortening_g"] + recipe["oil_g"] == 0:
        warnings.append("No butter, shortening, or oil quantity could be parsed.")
    if recipe["white_sugar_g"] + recipe["light_brown_sugar_g"] == 0:
        warnings.append("No white or brown sugar quantity could be parsed.")
    if recipe["chocolate_g"] == 0:
        warnings.append("No chocolate quantity was found; the prediction describes the base cookie dough.")
    if recipe["butter_g"] > 0 and not butter_state_found:
        warnings.append("No butter state was found; defaulted to softened butter.")
    if not mixing_method_found:
        warnings.append(f"No mixing method was found; defaulted to {recipe['mixing_method']} mixing.")
    if not chill_found:
        warnings.append("No chill instruction was found; defaulted to 0 hours.")
    if not temperature_found:
        warnings.append("No bake temperature was found; defaulted to 350°F.")
    if not bake_time_found:
        warnings.append("No bake time was found; defaulted to 10 minutes.")
    if not cookie_size_found:
        warnings.append("No cookie size was found; defaulted to 30 g.")
    if unparsed_lines:
        warnings.append(
            "Could not map these measured ingredient lines: " + "; ".join(unparsed_lines[:4])
        )

    return ParsedCookieRecipe(
        recipe=recipe,
        warnings=warnings,
        unparsed_lines=unparsed_lines,
    )
